File: common/src/main/java/generations/gg/generations/core/generationscore/world/level/block/entities/SimpleBlockEntity.java
package generations.gg.generations.core.generationscore.world.level.block.entities;

import com.pokemod.pokemod.api.data.Syncable;
import com.pokemod.pokemod.client.model.ModelContextProviders;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.protocol.Packet;
import net.minecraft.network.protocol.game.ClientGamePacketListener;
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
import net.minecraft.world.level.block.HorizontalDirectionalBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class SimpleBlockEntity extends BlockEntity implements Syncable, ModelContextProviders.AngleProvider {

    public SimpleBlockEntity(BlockEntityType<?> pType, BlockPos pPos, BlockState pBlockState) {
        super(pType, pPos, pBlockState);
    }

    protected void readNbt(CompoundTag nbt) {}
    protected void writeNbt(CompoundTag nbt) {}

    public void sync() {
        setChanged();
        if(getLevel() != null)
            getLevel().sendBlockUpdated(getBlockPos(), getBlockState(), getBlockState(), 2);
    }

    @Override
    public void load(@NotNull CompoundTag nbt) {
        super.load(nbt);
        readNbt(nbt);
    }

    @Override
    protected void saveAdditional(@NotNull CompoundTag nbt) {
        super.saveAdditional(nbt);
        writeNbt(nbt);
    }

    @Override
    public @NotNull CompoundTag getUpdateTag() {
        CompoundTag tag = new CompoundTag();
        saveAdditional(tag);
        return tag;
    }

    @Nullable
    @Override
    public Packet<ClientGamePacketListener> getUpdatePacket() {
        return ClientboundBlockEntityDataPacket.create(this);
    }

    @Override
    public float getAngle() {
        if (getBlockState().hasProperty(HorizontalDirectionalBlock.FACING))
            return getBlockState().getValue(HorizontalDirectionalBlock.FACING).toYRot();
        return 0f;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/level/block/entities/SimpleBlockEntity.kt
package generations.gg.generations.core.generationscore.common.world.level.block.entities

import generations.gg.generations.core.generationscore.common.client.model.ModelContextProviders.AngleProvider
import net.minecraft.core.BlockPos
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.protocol.Packet
import net.minecraft.network.protocol.game.ClientGamePacketListener
import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket
import net.minecraft.world.level.block.HorizontalDirectionalBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState

abstract class SimpleBlockEntity(pType: BlockEntityType<*>, pPos: BlockPos, pBlockState: BlockState) :
    BlockEntity(pType, pPos, pBlockState), AngleProvider {
    protected open fun readNbt(nbt: CompoundTag, provider: HolderLookup.Provider) {}
    protected open fun writeNbt(nbt: CompoundTag, provider: HolderLookup.Provider) {}

    fun sync() {
        setChanged()
        level?.sendBlockUpdated(blockPos, blockState, blockState, 2)
    }

    override fun loadAdditional(nbt: CompoundTag, provider: HolderLookup.Provider) {
        super.loadAdditional(nbt, provider)
        readNbt(nbt, provider)
    }

    override fun saveAdditional(nbt: CompoundTag, provider: HolderLookup.Provider) {
        super.saveAdditional(nbt, provider)
        writeNbt(nbt, provider)
    }

    override fun getUpdateTag(registries: HolderLookup.Provider): CompoundTag {
        val tag = CompoundTag()
        writeNbt(tag, registries)
        return tag
    }

    override fun getUpdatePacket(): Packet<ClientGamePacketListener>? {
        return ClientboundBlockEntityDataPacket.create(this)
    }

    override fun getAngle(): Float {
        if (blockState.hasProperty(HorizontalDirectionalBlock.FACING)) return blockState.getValue(
            HorizontalDirectionalBlock.FACING
        ).toYRot()
        return 0f
    }
}